python/cudf/cudf/core/spill_manager.py

```diff
@@ -0,0 +1,368 @@
+                    break  # No more to spill!
+                total_spilled += spilled
+
+            if total_spilled > 0:
+                return True  # Ask RMM to retry the allocation
+
+            if retry_on_error:
+                # Let's collect garbage and try one more time
+                gc.collect()
+                return oom(nbytes, retry_on_error=False)
+
+            # TODO: write to log instead of stdout
+            print(
+                f"[WARNING] RMM allocation of {format_bytes(nbytes)} bytes "
+                "failed, spill-on-demand couldn't find any device memory to "
+                f"spill:\n{repr(self)}\ntraceback:\n{get_traceback()}"
+            )
+            if self._expose_statistics is None:
+                print("Set `CUDF_SPILL_STAT_EXPOSE=on` for expose statistics")
+            else:
+                print(self.pprint_expose_statistics())
+
+            return False  # Since we didn't find anything to spill, we give up
+
+        current_mr = rmm.mr.get_current_device_resource()
+        mr = rmm.mr.FailureCallbackResourceAdaptor(current_mr, oom)
+        rmm.mr.set_current_device_resource(mr)
+
+    def add(self, buffer: SpillableBuffer) -> None:
+        if buffer.size > 0 and not buffer.exposed:
+            with self._lock:
+                self._base_buffers[self._id_counter] = buffer
+                self._id_counter += 1
+
+    def add_other(self, buffer: DeviceBufferLike) -> None:
+        with self._lock:
+            if buffer.size > 0:
+                self._other_buffers[self._id_counter] = buffer
+                self._id_counter += 1
+
+    def base_buffers(
+        self, order_by_access_time: bool = False
+    ) -> Tuple[SpillableBuffer, ...]:
+        with self._lock:
+            ret = tuple(self._base_buffers.values())
+        if order_by_access_time:
+            ret = tuple(sorted(ret, key=lambda b: b.last_accessed))
+        return ret
+
+    def other_buffers(self) -> Tuple[DeviceBufferLike, ...]:
+        with self._lock:
+            ret = tuple(self._other_buffers.values())
+        return ret
+
+    def spilled_and_unspilled(self) -> Tuple[int, int]:
+        spilled, unspilled = 0, 0
+        for buf in self.base_buffers():
+            if buf.is_spilled:
+                spilled += buf.size
+            else:
+                unspilled += buf.size
+        return spilled, unspilled
+
+    def spill_device_memory(self) -> int:
+        """Try to spill device memory
+
+        This function is safe to call doing spill-on-demand
+        since it does not lock buffers already locked.
+
+        Return
+        ------
+        int
+            Number of bytes spilled.
+        """
+        for buf in self.base_buffers(order_by_access_time=True):
+            if buf.lock.acquire(blocking=False):
+                try:
+                    if not buf.is_spilled and buf.spillable:
+                        buf.__spill__(target="cpu")
+                        return buf.size
```

Comment:
It seems odd that calling code needs to loop and call this method when this method itself is going to loop. Is it required because the calling methods all need more control over how much is spilled?

Reply:
This is looping to handle the case where `buf.lock.acquire(blocking=False) is False`.